### PR TITLE
Bump numpy version >=1.17.0

### PR DESCRIPTION
**Pull request by reminisce**
## Description ##
To adopt numpy `__array_function__` protocol for better `ndarray` interoperability, we need numpy >= 1.17.0. See discussion in https://github.com/apache/incubator-mxnet/pull/16097.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by szha** [comment]
@leezu FYI

File: python/setup.py [hunk 1]
```diff
@@ -30,7 +30,11 @@
 else:
     from setuptools import setup
     from setuptools.extension import Extension
-    kwargs = {'install_requires': ['numpy>=1.17.0,<2.0.0', 'requests>=2.20.0,<3', 'graphviz<0.9.0,>=0.8.1'], 'zip_safe': False}
+    kwargs = {
+        'install_requires': ['requests>=2.20.0,<3', 'graphviz<0.9.0,>=0.8.1']
+            .append('numpy>=1.17.0,<2.0.0' if sys.version_info[0] > 2 else 'numpy>1.16.0,<2.0.0'),
```
Comment (leezu) [inline python/setup.py]: Why not use https://www.python.org/dev/peps/pep-0508/#environment-markers ?

**Comment by leezu** [comment]
`numba<0.45.1` does not work with numpy 1.17. So users of numba must upgrade when using numpy 1.17. I believe we should nevertheless bump the MXNet numpy dependency version. But FYI, to make you aware of the caveat.

File: python/setup.py (hunk 1 above)
Comment (reminisce) [reply python/setup.py]: Thanks for pointing this out. I'm not familiar with the tool, but will adopt it in this PR after CI passes.

File: python/setup.py [hunk 2]
```diff
@@ -30,7 +30,12 @@
 else:
     from setuptools import setup
     from setuptools.extension import Extension
-    kwargs = {'install_requires': ['numpy>1.16.0,<2.0.0', 'requests>=2.20.0,<3', 'graphviz<0.9.0,>=0.8.1'], 'zip_safe': False}
+    kwargs = {
+        'install_requires': ['requests>=2.20.0,<3', 'graphviz<0.9.0,>=0.8.1']
+            .append('numpy>=1.17.0,<2.0.0' if sys.version_info.major >= 3 and sys.version_info.minor >= 5
+                    else 'numpy>1.16.0,<2.0.0'),
```
Comment (szha) [inline python/setup.py]: what's the concern for just switching to `>= 1.17`? was it CI?
setup.py may be evaluated on a platform that builds binary distributions that is different from the platform on which the distribution is installed. in this scenario, the condition won't be evaluated on the target platform

Comment (leezu) [reply python/setup.py]: `1.17` is Python 3 only.
If the condition is not evaluated on the target platform, we can't bump the version without dropping Python 2 support..

Comment (szha) [reply python/setup.py]: agreed. let's put this change on hold

**Pull request closed without merging**